cdispyutils/profiling/vis.py

```diff
@@ -0,0 +1,173 @@
+from functools import reduce
+import os
+import pstats
+
+import matplotlib.pyplot as plt
+from matplotlib.backends.backend_pdf import PdfPages
+import numpy as np
+
+
+class StatsCollection(object):
+
+    def __init__(self, directory):
+        self.stats = {}
+        for path, _, files in os.walk(directory):
+            files = [f for f in files if f.endswith(".prof")]
+            split = path.split(os.sep)
+            here = reduce(lambda acc, p: acc.get(p, {}), split[:-1], self.stats)
+            print(path)
+            print(files)
+            if files:
+                print(os.path.join(path, files[0]))
+            here[split[-1]] = {f: pstats.Stats(os.path.join(path, f)) for f in files}
+        self.stats = self.stats[directory]
+
+
+class ProfilePlotter(object):
+    """
+    Generate plots loaded from some profiling output previously recorded by a
+    ``cdispyutils.profiling.Profiler``.
+    """
+
+    def __init__(self, directory="profile"):
+        self.stats = StatsCollection(directory)
+
+    def make_all_plots(self, save_file="profile_graphs.pdf"):
+        """
+        Using the stats loaded into the ``StatsCollection``, generate a series of plots
+        for all the profiling files which are recorded, comparing the results from
+        different runs for their shared profiling events.
+
+        The stats in the ``StatsCollection`` have to be slightly reorganized for a
+        couple of reasons:
+        - We want to parametrize over separate runs, for every category
+        - For WSGI's profiling output, we want to aggregate the results for each run
+          into a single data point with an uncertainty value
+
+        The ``results`` variable will look like this:
+
+            {
+                "init": {
+                    "app_register_blueprints.prof": {
+                        "run": ["2018-11-30T15:15:36.14", "2018-11-30T15:15:38.21"],
+                        "result": [0.03, 0.04],
+                    },
+                },
+                "wsgi": {
+                    "GET.root.prof": {
+                        "run": ["2018-11-30T15:15:36.14", "2018-11-30T15:15:38.21"],
+                        "result": [[0.0001, 0.0002, 0.0004], [0.0001]],
+                    },
+                    "GET._status.prof": {
+                        "run": ["2018-11-30T15:15:36.14", "2018-11-30T15:15:38.21"],
+                        "result": [[0.002], [0.004]],
+                    }
+                },
+            }
+
+        Then we assemble plots for every profile in every category, where the data
+        points are the run name as the x and the result time as the y.
+        """
+        results = {}
+        for run, categories in self.stats.stats.iteritems():
+            for category, files in categories.iteritems():
+                if category not in results:
+                    results[category] = {}
+                if category == "wsgi":
```

Comment:
this if/else for the category (which happens below too) is a bit of a code smell. It might be better to break out a class per category or at least one function to call to reduce code duplication as much as possible? it's a small complaint and I don't have a great solution off the top of my head but it seems like there's gotta be a way to be more DRY with this. {{Insert other buzzwords here}}